ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -113,19 +136,41 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
   if (TraceIdx == numeric_limits<uint64_t>::max())
     errx(EXIT_FAILURE, "trace index counter overflowed");
 
+  auto SrcMod = load_module(Context, SecPtr, SecSize);
+
+  // Get var args from start_tracing call.
+  auto Inputs = get_trace_inputs(SrcMod->getFunction(FuncNames[0]), BBs[0]);
+
+  std::vector<Type*> InputTypes;
+  for (auto Val: Inputs) {
+    InputTypes.push_back(Val->getType());
+  }
+
+  // Create function to store compiled trace.
   string TraceName = string(TRACE_FUNC_PREFIX) + to_string(TraceIdx);
   llvm::FunctionType *FType =
-      llvm::FunctionType::get(Type::getVoidTy(Context), false);
+      llvm::FunctionType::get(Type::getVoidTy(Context), InputTypes, false);
   llvm::Function *DstFunc = llvm::Function::Create(
       FType, Function::InternalLinkage, TraceName, DstMod);
   DstFunc->setCallingConv(CallingConv::C);
+
+  // Create entry block and setup builder.
   auto DstBB = BasicBlock::Create(Context, "", DstFunc);
   llvm::IRBuilder<> Builder(Context);
   Builder.SetInsertPoint(DstBB);
 
-  auto SrcMod = load_module(Context, SecPtr, SecSize);
-
   llvm::ValueToValueMapTy VMap;
+  // Map trace inputs to their appropriates variables in the trace.
+  for (size_t Idx = 0; Idx != Inputs.size(); Idx++) {
+    Value* OldVal = Inputs[Idx];
```

Comment:
In C (I don't know about C++) the convention is that the `*` goes next to the variable name so `Value *OldVal` to avoid the problem that `X* y, z` defines `y` as type `X*` and `z` as `X` (seriously!).

Reply:
I prefer putting the star next to the variable name (not the type) too.

Reply:
Whoops. That's a typo. Although this is the first time I hear a good reason why this is the standard, as it previously didn't make any sense to me (semantically the pointer is part of the type).

Reply:
Fixed in 18b0420.